Sources/X509/Verifier/OneOfPolicies.swift

```diff
@@ -0,0 +1,175 @@
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftCertificates project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftCertificates project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftASN1
+
+/// Provides a result-builder style DSL for constructing a ``VerifierPolicy`` in which one of the specified policies must match.
+///
+/// This DSL allows us to construct ``OneOfPolicies`` within a ``PolicyBuilder``.
+/// ```swift
+/// let verifier = Verifier(rootCertificates: CertificateStore()) {
+///     RFC5280Policy(validationTime: Date())
+///     OneOfPolicies {
+///         PolicyA()
+///         PolicyB()
+///     }
+/// }
+/// ```
+@resultBuilder
+public struct OneOfPolicyBuilder {}
+
+extension OneOfPolicyBuilder {
+    @inlinable
+    public static func buildLimitedAvailability<Policy: VerifierPolicy>(_ component: Policy) -> Policy {
+        component
+    }
+}
+
+// MARK: empty policy
+extension OneOfPolicyBuilder {
+    @usableFromInline
+    struct Empty: VerifierPolicy {
+        @inlinable
+        var verifyingCriticalExtensions: [SwiftASN1.ASN1ObjectIdentifier] { [] }
+
+        @inlinable
+        init() {}
+
+        @inlinable
+        mutating func chainMeetsPolicyRequirements(chain: UnverifiedCertificateChain) async -> PolicyEvaluationResult {
+            .failsToMeetPolicy(reason: "No policies specified in OneOfPolicies block")
+        }
+    }
+
+    @inlinable
+    public static func buildBlock() -> some VerifierPolicy {
+        Empty()
+    }
+}
+
+// MARK: concatenated policies
+extension OneOfPolicyBuilder {
+    @usableFromInline
+    struct Tuple2<First: VerifierPolicy, Second: VerifierPolicy>: VerifierPolicy {
+        @usableFromInline
+        var first: First
+
+        @usableFromInline
+        var second: Second
+
+        @inlinable
+        init(first: First, second: Second) {
+            self.first = first
+            self.second = second
+        }
+
+        @inlinable
+        var verifyingCriticalExtensions: [SwiftASN1.ASN1ObjectIdentifier] {
+            let firstExtensions = first.verifyingCriticalExtensions
+            let secondExtensions = second.verifyingCriticalExtensions
+            return firstExtensions.filter { secondExtensions.contains($0) }
+        }
+
+        @inlinable
+        mutating func chainMeetsPolicyRequirements(chain: UnverifiedCertificateChain) async -> PolicyEvaluationResult {
+            let firstResult = await self.first.chainMeetsPolicyRequirements(chain: chain)
+            switch firstResult {
+            case .meetsPolicy:
+                return .meetsPolicy
+            case .failsToMeetPolicy(let firstReason):
+                let secondResult = await self.second.chainMeetsPolicyRequirements(chain: chain)
```

Comment:
hm... it would be great if we could log a diagnostics here but we don't have access to the `diagnosticsHandler` given to `verify`. We could add a new method with the log handler as an "optional" requirement  to the `VerifierPolicy` protocol in an API stable way. Not great but possible. Do you have thoughts on this @Lukasa?

Reply:
To be concrete, I'm thinking about something like this:
```swift
extension VerifierPolicy {
    ...

    mutating func chainMeetsPolicyRequirements(chain: UnverifiedCertificateChain, logger: VerifierLogger) async -> PolicyEvaluationResult
}

extension VerifierPolicy {
    public mutating func chainMeetsPolicyRequirements(chain: UnverifiedCertificateChain, logger: VerifierLogger) async -> PolicyEvaluationResult {
        await self.chainMeetsPolicyRequirements(chain: chain)
    }
}

public struct VerifierLogger {
    var _log: Optional<(VerificationDiagnostic) -> ()>
    internal func log(_ diagnostics: @autoclosure @Sendable () -> VerificationDiagnostic) {
        _log?(diagnostics())
    }
    public func log(_ diagnostics: @autoclosure @Sendable () -> String) {
        _log?(/* user defined diagnostics */)
    }
}
```

Reply:
This is a good idea. However, considering the size of this PR already, do you think we could do this in a followup?

Reply:
Sure